File: src/main/java/com/emmanuelirem/studentassistant/models/Lecturer.java
package com.emmanuelirem.studentassistant.models;

import com.emmanuelirem.studentassistant.models.helper.MatchesIdPattern;
import com.emmanuelirem.studentassistant.models.university.Department;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Document
public class Lecturer {

    @Id
    private String id = UUID.randomUUID().toString();
    private String firstName;
    private String lastName;
    @JsonProperty("email")
    private String schoolEmailAddress;
    private String office;

    @JsonProperty("phoneNumbers")
    private ArrayList<String> personalPhoneNumber;

    @MatchesIdPattern
    private String username;

    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    private String password;

    @DBRef
    private List<Department> departments = new ArrayList<>();

    @DBRef
    private List<Course> courses = new ArrayList<>();

    @DBRef(lazy = true)
    @JsonIgnore
    private List<Message> messages = new ArrayList<>();

    private boolean inOffice = false;

    public Lecturer() {
    }

    public Lecturer(String firstName, String lastName, String username, String password, String office) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.username = username;
        this.password = password;
        this.office = office;
    }

    public Lecturer(String firstName, String lastName, String username, String password, List<Course> courses) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.username = username;
        this.password = password;
        this.courses.addAll(courses);
    }

    public Lecturer(String firstName, String lastName, ArrayList<String> personalPhoneNumber, String username, String password) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.personalPhoneNumber = personalPhoneNumber;
        this.username = username;
        this.password = password;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getSchoolEmailAddress() {
        return schoolEmailAddress;
    }

    public void setSchoolEmailAddress(String schoolEmailAddress) {
        this.schoolEmailAddress = schoolEmailAddress;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public List<Course> getCourses() {
        return courses;
    }

    public List<Department> getDepartments() {
        return departments;
    }

    public void setDepartments(List<Department> departments) {
        this.departments = departments;
    }

    public void addDepartment(Department department){
        if(!this.departments.contains(department))
            departments.add(department);
    }

    public void removeDepartment(Department department){
        if(this.departments.contains(department))
            departments.remove(department);
    }

    public String getOffice() {
        return office;
    }

    public void setOffice(String office) {
        this.office = office;
    }

    public void addCourse(Course course) {
        if(courses == null) {
            courses = new ArrayList<>();
        }

        if(!courses.contains(course)){
            courses.add(course);
            course.addLecturer(this);
        }
    }

    public void removeCourse(Course course) {
            courses.remove(course);
            course.removeLecturer(this);
    }

    public void setCourses(List<Course> courses) {
        if(!courses.isEmpty()){
            courses.forEach(this::addCourse);
        }
    }

    public List<Message> getMessages() {
        return messages;
    }

    public void setMessages(List<Message> messages) {
        this.messages = messages;
    }

    public void addMessage(Message message) {
        messages.add(message);
    }

    public void removeMessage(Message message){
            messages.remove(message);
    }

    public ArrayList<String> getPersonalPhoneNumber() {
        return personalPhoneNumber;
    }

    public void setPersonalPhoneNumber(ArrayList<String> personalPhoneNumber) {
        this.personalPhoneNumber = personalPhoneNumber;
    }

    public void addPhone(String number) {
        if(!personalPhoneNumber.contains(number))
            personalPhoneNumber.add(number);
    }

    public boolean isInOffice() {
        return inOffice;
    }

    public void setInOffice(boolean inOffice) {
        this.inOffice = inOffice;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Lecturer lecturer = (Lecturer) o;

        if (!id.equals(lecturer.id)) return false;
        if (firstName != null ? !firstName.equals(lecturer.firstName) : lecturer.firstName != null) return false;
        if (lastName != null ? !lastName.equals(lecturer.lastName) : lecturer.lastName != null) return false;
        return username.equals(lecturer.username);
    }

    @Override
    public int hashCode() {
        int result = id.hashCode();
        result = 31 * result + (firstName != null ? firstName.hashCode() : 0);
        result = 31 * result + (lastName != null ? lastName.hashCode() : 0);
        result = 31 * result + username.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return "Lecturer{" +
                "id=" + id +
                ", firstName='" + firstName + '\'' +
                ", lastName='" + lastName + '\'' +
                ", schoolEmailAddress='" + schoolEmailAddress + '\'' +
                ", office='" + office + '\'' +
                ", username='" + username + '\'' +
                ", password='" + password + '\'' +
                '}';
    }
}

File: client/src/model/Lecturer.ts
import Department from '@/model/Department';
import Course from '@/model/Course';

export default class Lecturer {
    private _id: string = '';
    private _firstName: string = '';
    private _lastName: string = '';
    private _email: string = '';
    private _office: string = '';
    private _inOffice: boolean = false;
    private _phoneNumbers: string[] = [];
    private _username: string = '';
    private _password: string = '';
    private _departments: Department[] = [];
    private _courses: Course[] = [];

    constructor() {
    }

    get id(): string {
        return this._id;
    }

    set id(value: string) {
        this._id = value;
    }

    get firstName(): string {
        return this._firstName;
    }

    set firstName(value: string) {
        this._firstName = value;
    }

    get lastName(): string {
        return this._lastName;
    }

    set lastName(value: string) {
        this._lastName = value;
    }

    get email(): string {
        return this._email;
    }

    set email(value: string) {
        this._email = value;
    }

    get office(): string {
        return this._office;
    }

    set office(value: string) {
        this._office = value;
    }

    get inOffice(): boolean {
        return this._inOffice;
    }

    set inOffice(value: boolean) {
        this._inOffice = value;
    }

    get phoneNumbers(): string[] {
        return this._phoneNumbers;
    }

    set phoneNumbers(value: string[]) {
        this._phoneNumbers = value;
    }

    get username(): string {
        return this._username;
    }

    set username(value: string) {
        this._username = value;
    }


    get password(): string {
        return this._password;
    }

    set password(value: string) {
        this._password = value;
    }

    get departments(): Department[] {
        return this._departments;
    }

    set departments(value: Department[]) {
        this._departments = value;
    }

    get courses(): Course[] {
        return this._courses;
    }

    set courses(value: Course[]) {
        this._courses = value;
    }

    public addDepartment(department: Department): Department[] {
        this.departments.push(department);
        return this.departments;
    }

    public removeDepartment(departmentId: string): Department[] {
        const id = this.courses.findIndex((value) => value.id === departmentId);
        this.departments.splice(id, 1);
        return this.departments;
    }

    public addCourse(course: Course): Course[] {
        this.courses.push(course);
        return this.courses;
    }

    public removeCourse(courseId: string): Course[] {
        const id = this.courses.findIndex((value) => value.id === courseId);
        this.courses.splice(id, 1);
        return this.courses;
    }


    get json(): {} {
        return {
            firstName: this._firstName,
            lastName: this._lastName,
            email: this._email,
            office: this._office,
            inOffice: this._inOffice,
            phoneNumbers: this._phoneNumbers,
            username: this._username,
            password: this._password,
            departments: this._departments,
            courses: this._courses,
        };
    }
}
